### ISSUE TITLE
STREAMCMD in cg-proc

### ISSUE THREAD

**Issue by unhammer** [TinoDidriksen/cg3#68]
cg-proc doesn't seem to support STREAMCMD. It'd be nice to be able to do something like `[<STREAMCMD:SETVAR:coke=pop>]` so we can `REMOVE ("soda") IF (0 (VAR:coke=pop));` without having to compile extra autobil.bin's (and run extra pipelines) for every such preference.

It should be possible to send this in along with the formatted text. I'm not sure what the best way to do it would be, perhaps simply `<STREAMCMD:SETVAR:coke=pop><html><body>Quiero un refresco</body></html>`  should be possible?

**Comment by unhammer** [comment]
@khannatanmai thoughts? This relates to your work :)  cg-proc needs to see the `STREAMCMD:SETVAR:coke=pop` string, but the rest of the pipeline shouldn't, so it should be in a blank. I don't think it can be considered a word-blank, because you wouldn't want to duplicate it (consider a CG rule that, on seeing the variable set, unsets it thinking it will be unset until it actually appears in input again). 

Maybe if `tf-extract` produced

```
[transfuse:\/tmp\/transfuse-aHRUouiIE6s]
[STREAMCMD:SETVAR:coke=pop]
[tf-block:1-lhTrhg]

Quiero un refresco.[]
```
and tf-inject ignored any STREAMCMD?

**Comment by TinoDidriksen** [comment]
I don't see how it relates to khannatanmai's work or Transfuse. There is no inherent reason these variables are at the start of a stream, and they can be changed mid-stream if circumstances require it. While they are mostly used as domain information that is per-stream, this is just one use.

That said, Transfuse could be told to extract such variables from [XML processing instructions](https://en.wikipedia.org/wiki/Processing_Instruction). That wouldn't work for docx/odt or other binary-wrapped formats, though.

Best if it comes from a tool that injects it from environment or config file into the stream at some suitable point, probably immediately after morph analysis.

**Comment by unhammer** [comment]
I didn't say they had to be at the start of a stream, but that they shouldn't be duplicated (nor removed; ie. they can't be treated as word-blanks). 

It relates to transfuse in several ways, one of which you may have answered:
1. What is the best way to put it into the stream so that transfuse can pass it on (processing instructions then? no need for `<apertium-notrans>`?)
2. How should it appear in the stream so that cg-proc can read it (is simply `[<STREAMCMD:SETVAR:coke=pop>]` ok?)
3. Currently, transfuse does *not* let processing instructions appear in the stream, and segfaults if you just put it in there without any wrapping:
```
$ cat inputs.html
<html><apertium-notrans><STREAMCMD:SETVAR:coke=pop></apertium-notrans><body>Quiero un refresco</body></html>
$ tf-extract -f html inputs.html
Segmentation fault (core dumped)
```
similarly at the start of stream, or without notrans; while if I put it in a `<? ?>` wrapper it's hidden from the stream.

**Comment by TinoDidriksen** [comment]
There is currently no way to do any of this, so it's all hypothetical and implementation ideas.

I don't see how Transfuse could ever do this for formats beyond HTML. It wouldn't be possible to prepend data to a DOCX. APy doesn't keep formatted pipes open, though the non-Transfuse parts could and should be kept open - the Transfuse parts can't be. Overall, I don't think this belongs anywhere near Transfuse - format handling must be kept agnostic to such side-channel information.

A way for APy to do this would be an extra arg in the request that APy turns into Streamcmds between Transfuse and the translation pipe. That is also how GramTrans does it.

**Comment by unhammer** [comment]
I tried this. So the design is:

The stream command is inserted (and removed) between transfuse and the regular pipeline, like
```sh
$ tf-extract … | setvar_wrap | bash <(apertium-wblank-mode foo-bar.mode) … | setvar_unwrap | tf-inject …
```
(done by https://github.com/apertium/apertium/compare/master...unhammer:biprefs?expand=1 – the gawk scripts are slightly complicated by handling both NUL flushing and per-newline output so they don't sponge up the stream before outputting)

and then it's available in the stream for cg-proc (which required a patch https://github.com/TinoDidriksen/cg3/compare/master...unhammer:biprefs?expand=1 ). I made the format quite specific to make it easily parseable, so it has to be exactly `[<STREAMCMD:SETVAR:_>]` where only the _ can change and can't contain `>` or `]` and it has to start on the top level (not inside another blank).

I use it here https://github.com/apertium/apertium-nno-nob/compare/biprefs?expand=1 to select bilingual dictionary translations, e.g. 

```sh
$ export AP_SETVAR
$ unset AP_SETVAR; echo uenige i forskjellen| ~/PREFIX/lttoolbox/bin/apertium -d . nob-nno_e
usamde i forskjellen
$ export AP_SETVAR="forskjell_skilnad,samd_einig"
$ echo uenige i forskjellen| ~/PREFIX/lttoolbox/bin/apertium -d . nob-nno_e
ueinige i skilnaden
```


I haven't done the APy changes yet, but for long-running pipelines, APy would hold the programs in foo-bar.mode opened, while manually inserting and stripping the SETVAR's for every request.

Does this seem mergeable?
